### PR TITLE
Change how hashtags are normalized

### PR DESCRIPTION
**Pull request by Gargron**
Previously, `ｓｙｎｔｈｗａｖｅ` and `synthwave` were both allowed to be hashtags, but would be considered different hashtags. This PR introduces a normalization algorithm that goes beyond case insensitivity and mirrors what we use in the ElasticSearch index to ensure that hashtags can be found broadly:

- CJK width normalization (e.g. full-width Latin text, half-width Katakana, and so on)
- ASCII folding (e.g. turn `Å` into `A`)

Additionally to the previous normalization methods such as unicode-aware downcasing and stripping out characters that are not considered a valid part of the hashtag (previously, this was only a leading `#`, but we can expand that to just anything that doesn't match the original regex). This normalization goes into the previously existing `name` column on the `tags` table. However, a new column `display_name` is added to store the pre-normalized version to ensure that while `BLÅHAJ` results in a normalized `blahaj` value for retrieval purposes, the string `BLÅHAJ` is still displayed outward.

There are no API changes.

Additionally, I discovered that Ruby's `[:word:]` character class matches the unicode Mark class, which includes various invisible characters. I believe `[:alnum:]` is more in the spirit of what was intended.

Note: The PR does not make an effort to migrate any existing posts to new hashtags, or to deduplicate old hashtags. A data migration is not strictly necessary to improve future user experience.

### CODE REVIEW

**Review by ykzts — APPROVED** [review]

**Review by mashirozx — APPROVED** [review]

File: app/models/account.rb [hunk 1]
```diff
@@ -62,7 +62,7 @@ class Account < ApplicationRecord
   )
 
   USERNAME_RE   = /[a-z0-9_]+([a-z0-9_\.-]+[a-z0-9_]+)?/i
-  MENTION_RE    = /(?<=^|[^\/[:word:]])@((#{USERNAME_RE})(?:@[[:word:]\.\-]+[[:word:]]+)?)/i
+  MENTION_RE    = /(?<=^|[^\/[:word:]])@((#{USERNAME_RE})(?:@[[:alnum:]\.\-]+[[:alnum:]]+)?)/i
```
Comment (ClearlyClaire) [inline app/models/account.rb:65]: Are you sure `[[:alnum:]]` cover everything that is allowed in an International Domain Name?

**Comment by MarcusWolschon** [comment]
@Gargron  you made a mistake in using
https://github.com/mastodon/mastodon/blob/a3f176423ff124981a2cf309257d2344171288d3/app/lib/ascii_folding.rb#L5
I explained it in #21081 .
German Umlauts are transliterated wrong.
This leads to hashtags of different words with completely different meaning.

It should be
Ä=Ae
Ü=Ue
Ö=Oe
ẞ=ss
ä=ae
ü=ue
ö=oe

Isn't there an existing, maintained library for this instead of inventing your own?
Can anyone explain why hashtags are limited to the US-ASCII character set instead of UTF-8 in the first place?
After all all of UTF-8 can nowadays be represented in any database field, URL, HTML and Domain name.
Do hashtags work for all non-latin alphabets with this implementation.

**Comment by Gargron** [comment]
> Can anyone explain why hashtags are limited to the US-ASCII character set instead of UTF-8 in the first place?

They are not. You misunderstand the purpose of this change. Hashtags are unicode. But not everybody enters dialectics the same way. Simplest example is the IKEA shark, properly spelled `BLÅHAJ`. Not everyone can type `Å` however, so people will be spelling it `blahaj`, or they will misremember where the special letter goes and spell it `BLAHÅJ`. Obviously all of those hashtags are supposed to be the same hashtag. This is where ASCII-folding comes in, so posts for all 3 variants can appear in the same place. This is not a user-facing thing, it affects internal storage. The hashtag's "display variant" is still Unicode.

As for German umlauts, we don't really have information if the letters are German or not.

**Comment by MarcusWolschon** [comment]
I tried to read my way through the unicode and utf8 case folding and normalisatio  but that left me completely confused.
(It seems you need to know Kituanien and Greek to understand the differences Ruby offers.)
I guess the best way for Umlauts is to remove them from this list until a linguist comes along.

**Comment by sozialwelten** [comment]
As it is, hashtags containing German Umlaute are not working properly in several ways.

Apart from the search for them using the UI (a being equivalent to ä which it should not be), hashtags as links do not always return all of the expected search results. Sometimes even none at all (Screenshot). 

I also was under the impression that entering a hashtag as an URL results in different outcomes depending on the machines/browsers character settings but as @Gargron [mentioned](https://github.com/mastodon/mastodon/pull/18795#issuecomment-1396920483) this might not be the case. However, there definitely is some undesirable behaviour.

moeglichkeit001_20230214_102528

**Pull request merged**